### PR TITLE
[jit] Erase shape information from class types

### PR DESCRIPTION
**Pull request by driazati**
Differential Revision: [D16681944](https://our.internmc.facebook.com/intern/diff/16681944/)

### CODE REVIEW

File: aten/src/ATen/core/jit_type.h [hunk 1]
```diff
@@ -1440,6 +1432,10 @@ struct CAFFE2_API ClassType : public NamedType {
     methods_.push_back(method);
   }
 
+  TypePtr withContained(std::vector<TypePtr> contained_types) {
+    return refine(std::move(contained_types));
```
Comment (eellison) [inline aten/src/ATen/core/jit_type.h]: This makes the assumption that the contained_types are all subtypes of the original type.

If one of the attributes has been set to DimensionedTensorType, then is being set to TensorType (as with unshapedType) then TensorType won't subtype DimensionedTensorType

Comment (eellison) [reply aten/src/ATen/core/jit_type.h]: Maybe i don't understand the use case as well

File: aten/src/ATen/core/jit_type.h [hunk 2]
```diff
@@ -1598,6 +1598,15 @@ struct CAFFE2_API ClassType : public NamedType {
   // These variants are not registered in the global class table.
   ClassTypePtr refine(at::ArrayRef<TypePtr> refined_slots) const;
 
+  TypePtr createWithContained(std::vector<TypePtr> contained_types) const override {
+    auto ptr = ClassType::create(name_, compilation_unit_);
+    AT_ASSERT(numAttributes() == contained_types.size());
+    for(size_t i = 0; i < attributeNames_.size(); ++i) {
+      ptr->addAttribute(attributeNames_[i], contained_types[i]);
```
Comment (eellison) [inline aten/src/ATen/core/jit_type.h]: Could we do any validation on the types here ? One should subtype the other or vice versa right ?

**Review by eellison — APPROVED** [review]
Can you add some test validation that the types were erased ? The current test doesn't throw an exception but that's all it's validating.

Maybe a c++ test where you create a class type with a Dimensioned Tensor Type and then set it to Tensor ?

File: test/test_jit.py [hunk 3]
```diff
@@ -13597,6 +13597,40 @@ def fn(x):
...
+                return torch.ops.quantized.fbgemm_linear_unpack(self._packed_weight)
+
+            @weight.setter
+            def weight(self, w):
+                self._packed_weight = torch.ops.quantized.fbgemm_linear_prepack(w)
+
+        with torch.jit._disable_emit_hooks():
+            x = torch.jit.script(Linear(10, 10))
+            torch._C._jit_pass_erase_shape_information(x.graph)
+
```
Comment (raghuramank100) [inline test/test_jit.py]: Can you add a test to verify that serialization (save and load) works for this module?

**Review by raghuramank100 — CHANGES_REQUESTED** [review]
Request test to verify serialization for modules containing packed weight tensors.

File: test/test_jit.py (hunk 3 above)
Comment (jamesr66a) [reply test/test_jit.py]: @raghuramank100  I'm going to test this as part of this change https://github.com/pytorch/pytorch/pull/23891

**Review by jamesr66a — APPROVED** [review]

File: aten/src/ATen/core/jit_type.h [hunk 4]
```diff
@@ -1627,6 +1627,16 @@ struct CAFFE2_API ClassType : public NamedType {
   // These variants are not registered in the global class table.
   ClassTypePtr refine(at::ArrayRef<TypePtr> refined_slots) const;
 
+  TypePtr createWithContained(std::vector<TypePtr> contained_types) const override {
+    auto ptr = ClassType::create(name_, compilation_unit_);
```
Comment (suo) [inline aten/src/ATen/core/jit_type.h]: @driazati: Can you add a comment explaining the difference between between this method and `refine`? They seem identical accept with the subtyping test swapped, which is a pretty subtle distinction.

**Pull request closed without merging**